test/hotspot/jtreg/compiler/lib/verify/Verify.java

```diff
@@ -71,107 +109,176 @@ private static void checkEQ(Object a, Object b, String context) {
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Value mismatch: " + a + " vs " + b);
         }
     }
 
     /**
      * Verify that two longs are identical.
      */
-    private static void checkEQimpl(long a, long b, String context) {
+    private void checkEQimpl(long a, long b, String field, Object aParent, Object bParent) {
         if (a != b) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b + " for " + context);
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch: " + a + " vs " + b);
+            print(a, b, field, aParent, bParent);
             throw new VerifyException("Value mismatch: " + a + " vs " + b);
         }
     }
 
     /**
-     * Verify that two floats have identical bits.
+     * There are two comparison modes: one where we compare the raw bits, which sees different NaN
+     * encodings as different values, and one where we see all NaN encodings as identical.
+     * Ideally, we would want to assert that the Float.floatToRawIntBits are identical.
+     * But the Java spec allows us to return different bits for a NaN, which allows swapping the inputs
+     * of an add or mul (NaN1 * NaN2 does not have same bits as NaN2 * NaN1, because the multiplication
+     * of two NaN values should always return the first of the two).
+     * Hence, by default, we pick the non-raw coparison: we verify that we have the same bit
+     * pattern in all cases, except for NaN we project to the canonical NaN, using Float.floatToIntBits.
      */
-    private static void checkEQimpl(float a, float b, String context) {
-        if (Float.floatToRawIntBits(a) != Float.floatToRawIntBits(b)) {
-            System.err.println("ERROR: Verify.checkEQ failed: value mismatch for " + context);
-            System.err.println("       Values: " + a + " vs " + b);
-            System.err.println("       Values: " + Float.floatToRawIntBits(a) + " vs " + Float.floatToRawIntBits(b));
+    private boolean isFloatEQ(float a, float b) {
+        return isFloatCheckWithRawBits ? Float.floatToRawIntBits(a) != Float.floatToRawIntBits(b)
+                                       : Float.floatToIntBits(a) != Float.floatToIntBits(b);
+    }
+
+    /**
+     * See comments for "isFloatEQ".
+     */
+    private boolean isDoubleEQ(double a, double b) {
+        return isFloatCheckWithRawBits ? Double.doubleToRawLongBits(a) != Double.doubleToRawLongBits(b)
+                                       : Double.doubleToLongBits(a) != Double.doubleToLongBits(b);
+    }
+
+    /**
+     * Check that two floats are equal according to "isFloatEQ".
+     */
+    private void checkEQimpl(float a, float b, String field, Object aParent, Object bParent) {
+        if (isFloatEQ(a, b)) {
+            System.err.println("ERROR: Verify.checkEQ failed: value mismatch. check raw: " + isFloatCheckWithRawBits);
```

Comment:
Just noticed this now (there are other places as well): Since we now have `Verify.checkEQ()` and `Verify.checkEQWithRawBits()`, it would improve the readability if we reported which method was used. It could be done with something like that (pseudo code):
```Java
System.err.println("ERROR: Verify.checkEQ" + withRawBitsString() + " failed: value mismatch.

String withRawBitsString() {
    return isFloatCheckWithRawBits ? "WithRawBits" : "";
}
```

Reply:
Boah. That is really going to bloat the code, don't you think?
The exception that is thrown will already give you the complete stack trace, including which methods were called. Is that not good enough?

Reply:
Hm, it could indeed be a little bit more complicated when you are deep down in a recursion. My thought was that it could be misleading when a test is using a mix of `verifyEQ()` and `verifyEQWithRawBits()` and you only read `verifyEQ` failed. You could be start looking at the wrong check even though the stack trace would have guided you to the correct place. Maybe we can just update "Verify.checkEQ" into something more generic like "Equality matching failed" and we're good. What do you think?